# Cars Dataset Cleaning and Preprocessing

In [1]:
import os
os.getcwd()

'C:\\Users\\jhkan\\Documents\\Projects\\Jaehong-Kang-Project-Repository\\Australian Used Car Price Prediction and Market Analysis\\python'

In [2]:
os.chdir("..")
os.getcwd()

'C:\\Users\\jhkan\\Documents\\Projects\\Jaehong-Kang-Project-Repository\\Australian Used Car Price Prediction and Market Analysis'

### Import

In [3]:
import pandas as pd

cars = pd.read_parquet("data/processed/cars_merged.parquet")

print(cars.shape)
print(cars.columns)
print(cars.info())
cars.head()

(53655, 22)
Index(['source', 'listing_id', 'title', 'make', 'model', 'year', 'kms',
       'price', 'body_type', 'transmission', 'fuel_type', 'engine_l',
       'drive_type', 'location', 'state', 'used_or_new', 'status', 'variant',
       'series', 'cc', 'color', 'seats'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53655 entries, 0 to 53654
Data columns (total 22 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   source        53655 non-null  object 
 1   listing_id    17043 non-null  float64
 2   title         33068 non-null  object 
 3   make          53655 non-null  object 
 4   model         53655 non-null  object 
 5   year          53655 non-null  int32  
 6   kms           53655 non-null  int64  
 7   price         53655 non-null  float64
 8   body_type     53655 non-null  object 
 9   transmission  53655 non-null  object 
 10  fuel_type     53651 non-null  object 
 11  engine_l      51213 non-null  f

,source,listing_id,title,make,model,year,kms,price,body_type,transmission,...,drive_type,location,state,used_or_new,status,variant,series,cc,color,seats
0,raw_1,NaN,None,alfa,romeo 4c,2016,27011,97644.0,Convertible,Sports Automatic Dual Clutch,...,None,Queensland,None,Used,None,None,None,NaN,None,NaN
1,raw_1,NaN,None,alfa,romeo stelvio,2023,3909,85400.0,SUV,Sports Automatic,...,None,Victoria,None,Demo,None,None,None,NaN,None,NaN
2,raw_1,NaN,None,alfa,romeo tonale,2023,22,65250.0,SUV,Sports Automatic Dual Clutch,...,None,Victoria,None,Demo,None,None,None,NaN,None,NaN
3,raw_1,NaN,None,alfa,romeo tonale,2024,20,75990.0,SUV,Sports Automatic Dual Clutch,...,None,Victoria,None,New,None,None,None,NaN,None,NaN
4,raw_1,NaN,None,aston,martin rapide,2014,66993,145890.0,Hatchback,Sports Automatic,...,None,New South Wales,None,Used,None,None,None,NaN,None,NaN


### Missing Values

In [4]:
cars.isna().sum().sort_values(ascending=False)

state           37630
drive_type      37630
seats           36612
listing_id      36612
color           36612
series          36612
variant         36612
status          36612
cc              22206
title           20587
location        17043
engine_l         2442
fuel_type           4
transmission        0
body_type           0
used_or_new         0
price               0
kms                 0
year                0
model               0
make                0
source              0
dtype: int64

In [5]:
(cars.isna().sum() / len(cars) * 100).sort_values(ascending=False)

state           70.133259
drive_type      70.133259
seats           68.235952
listing_id      68.235952
color           68.235952
series          68.235952
variant         68.235952
status          68.235952
cc              41.386637
title           38.369211
location        31.764048
engine_l         4.551300
fuel_type        0.007455
transmission     0.000000
body_type        0.000000
used_or_new      0.000000
price            0.000000
kms              0.000000
year             0.000000
model            0.000000
make             0.000000
source           0.000000
dtype: float64

### Describing values

In [6]:
cars.describe()

,listing_id,year,kms,price,engine_l,cc,seats
count,1.704300e+04,53655.000000,53655.000000,53655.000000,51213.000000,31449.000000,17043.000000
mean,1.279026e+07,2017.819849,80606.710465,40025.898369,2.392029,1352.197272,5.116353
std,5.051347e+04,5.146029,74902.710210,28652.094984,0.871974,1398.876856,1.121382
min,1.153013e+07,1980.000000,1.000000,900.000000,0.700000,0.000000,2.000000
25%,1.275715e+07,2015.000000,16624.500000,22163.000000,1.991000,4.000000,5.000000
50%,1.280207e+07,2019.000000,65022.000000,33980.000000,2.179000,1498.000000,5.000000
75%,1.283132e+07,2022.000000,121722.500000,48990.000000,2.800000,2440.000000,5.000000
max,1.285246e+07,2025.000000,987475.000000,499990.000000,7.300000,7300.000000,14.000000


### Column Inspections

In [7]:
for col in cars.select_dtypes(include='object'):
    print(col, cars[col].nunique())

source 3
title 16942
make 78
model 1039
body_type 62
transmission 13
fuel_type 15
drive_type 5
location 626
state 8
used_or_new 9
status 3
variant 2361
series 2286
color 235


Note:
- 13 transmission types
- 62 body types
- 15 fuel types
- 5 drive types
- 626 locations
- 8 states
- 9 used or new
- 2361 variants
- 235 colours

### Fixing Category names

#### Transmission

In [8]:
cars['transmission'].value_counts()

Automatic                           31181
Sports Automatic                     9252
Manual                               5510
Constantly Variable Transmission     3495
Sports Automatic Dual Clutch         3280
Reduction Gear                        545
-                                     215
Sports Automatic Single Clutch        146
AWD                                    13
Front                                   8
Automated Manual Transmission           5
Rear                                    3
Seq. Manual Auto-Single Clutch          2
Name: transmission, dtype: int64

In [9]:
# Clean formatting
cars['transmission'] = cars['transmission'].str.lower().str.strip()

# Replace incorrect entries
cars['transmission'] = cars['transmission'].replace({
    '-': None,
    'awd': None,
    'front': None,
    'rear': None
})

# Standardize categories
transmission_map = {
    'automatic': 'automatic',
    'sports automatic': 'automatic',
    'sports automatic dual clutch': 'dual_clutch',
    'sports automatic single clutch': 'automated_manual',
    'automated manual transmission': 'automated_manual',
    'constantly variable transmission': 'cvt',
    'manual': 'manual',
    'reduction gear': 'ev'
}

cars['transmission'] = cars['transmission'].replace(transmission_map)

In [10]:
cars['transmission'].value_counts(dropna=False)

automatic                         40433
manual                             5510
cvt                                3495
dual_clutch                        3280
ev                                  545
None                                239
automated_manual                    151
seq. manual auto-single clutch        2
Name: transmission, dtype: int64

Transmission now contains missing values, we will handle this below

#### Body Type

In [11]:
cars['body_type'].value_counts()

SUV                 17903
Wagon                9202
Hatchback            7422
Sedan                5992
Ute                  2958
                    ...  
King Cab Pick Up        1
Softback                1
Panel Van               1
Crew Cab Van            1
Cab Chassis Tray        1
Name: body_type, Length: 62, dtype: int64

In [12]:
cars['body_type'].unique()

array(['Convertible', 'SUV', 'Hatchback', 'Sedan', 'Wagon', 'Coupe',
       'Ute', 'Van', 'Truck', 'UTE', 'N/A', 'People Mover',
       'Dual Cab Pick-up', 'Crew Cab Pickup', 'Double Cab Chassis',
       'Hardtop', 'Crew Cab Utility', 'Utility', 'Double Cab Utility',
       'Liftback', 'Dual Cab Utility', 'Cab Chassis',
       'Double Cab Pick Up', 'Dual Cab Chassis', 'Sportswagon', 'Pickup',
       'Sportback', 'Space Cab Chassis', 'Crew Cab Chassis',
       'Space Cab Utility', 'Fastback', 'Bus', 'X Cab Cab Chassis',
       'Super Cab Utility', 'Softback', 'Cabriolet', 'King Cab Pickup',
       'Club Cab Utility', 'Saloon', 'X Cab Pickup', 'Roadster',
       'Leaf Cab Chassis', 'Freestyle Cab Chassis', 'Softtop',
       'Space Cab Pickup', 'Coil Cab Chassis', 'X Cab Utility',
       'Super Cab Chassis', 'Troop Carrier', 'Coach', 'King Cab Pick Up',
       'Club Cab Chassis', 'King Cab Utility', 'Blind Van',
       'Club Cab Pickup', 'Crew Van', 'Freestyle Utility',
       'Super Cab 

In [13]:
# Standardize
cars['body_type'] = cars['body_type'].str.lower().str.strip()

body_map = {
    'ute': 'ute',
    'utility': 'ute',
    'dual cab pick-up': 'ute',
    'dual cab pickup': 'ute',
    'double cab pickup': 'ute',
    'double cab pick up': 'ute',
    'crew cab pickup': 'ute',
    'crew cab utility': 'ute',
    'dual cab utility': 'ute',
    'double cab utility': 'ute',
    'club cab pickup': 'ute',
    'king cab pickup': 'ute',
    'king cab pick up': 'ute',
    'super cab pickup': 'ute',
    'x cab pickup': 'ute',

    'cab chassis': 'truck',
    'dual cab chassis': 'truck',
    'double cab chassis': 'truck',
    'crew cab chassis': 'truck',
    'super cab chassis': 'truck',
    'space cab chassis': 'truck',
    'x cab cab chassis': 'truck',
    'leaf cab chassis': 'truck',
    'freestyle cab chassis': 'truck',
    'coil cab chassis': 'truck',
    'club cab chassis': 'truck',

    'panel van': 'van',
    'crew van': 'van',
    'blind van': 'van',
    'bus': 'van',

    'people mover': 'people_mover',

    'saloon': 'sedan',
    'fastback': 'sedan',

    'estate': 'wagon',
    'sportswagon': 'wagon',

    'sportback': 'hatchback',
    'liftback': 'hatchback',

    'cabriolet': 'convertible',
    'roadster': 'convertible',
    'softtop': 'convertible',

    'n/a': None
}

cars['body_type'] = cars['body_type'].replace(body_map)

# Handle rare cab types
cars.loc[cars['body_type'].str.contains('cab|pickup|utility', na=False), 'body_type'] = 'ute'
cars.loc[cars['body_type'].str.contains('van', na=False), 'body_type'] = 'van'

In [14]:
cars['body_type'].value_counts()

suv              17903
wagon             9349
ute               7789
hatchback         7505
sedan             6041
van               1622
coupe             1100
truck              750
convertible        321
people_mover        46
hardtop             36
troop carrier        6
coach                4
softback             1
Name: body_type, dtype: int64

#### Drive type

In [15]:
cars['drive_type'].value_counts()

Front    6717
4WD      3057
AWD      2962
Rear     2179
Other    1110
Name: drive_type, dtype: int64

In [16]:
# Standardize
cars['drive_type'] = cars['drive_type'].str.lower().str.strip()

drive_map = {
    'front': 'fwd',
    'rear': 'rwd',
    'awd': 'awd',
    '4wd': '4wd',
    'other': 'unknown',
    'unknown': 'unknown'
}

cars['drive_type'] = cars['drive_type'].replace(drive_map)

In [17]:
cars['drive_type'].value_counts(dropna=False)

None       37630
fwd         6717
4wd         3057
awd         2962
rwd         2179
unknown     1110
Name: drive_type, dtype: int64

There is a significant number of unknown drive types, will handle by associating make + model to body type

#### Fuel type

In [18]:
cars['fuel_type'].value_counts()

Diesel                       17322
Petrol                       11836
Unleaded Petrol               7015
Unleaded                      6847
Premium Unleaded Petrol       3434
Premium                       3230
Hybrid                        2077
Electric                       831
-                              534
Unleaded Petrol/Electric       255
Premium Unleaded/Electric      194
Liquid Petroleum Gas            40
Diesel/Electric                 18
LPG                             15
Leaded                           3
Name: fuel_type, dtype: int64

In [19]:
# Standarize
cars['fuel_type'] = cars['fuel_type'].str.lower().str.strip()

fuel_map = {
    'petrol': 'petrol',
    'unleaded petrol': 'petrol',
    'unleaded': 'petrol',
    'premium unleaded petrol': 'petrol',
    'premium': 'petrol',
    'leaded': 'petrol',

    'diesel': 'diesel',

    'hybrid': 'hybrid',
    'unleaded petrol/electric': 'hybrid',
    'premium unleaded/electric': 'hybrid',
    'diesel/electric': 'hybrid',

    'electric': 'electric',

    'liquid petroleum gas': 'lpg',
    'lpg': 'lpg',

    '-': None
}

cars['fuel_type'] = cars['fuel_type'].replace(fuel_map)

# Check
cars['fuel_type'].value_counts(dropna=False)

petrol      32365
diesel      17322
hybrid       2544
electric      831
None          538
lpg            55
Name: fuel_type, dtype: int64

#### Used or New

In [20]:
cars['used_or_new'].value_counts()

 Used            16299
USED             14961
Used             14454
New               3071
Demo              3062
NEW                568
DEMO               496
 Demo              387
 New In Stock      357
Name: used_or_new, dtype: int64

In [21]:
# Standarize
cars['used_or_new'] = (
    cars['used_or_new']
    .str.strip()
    .str.lower()
)

condition_map = {
    'new in stock': 'new'
}

cars['used_or_new'] = cars['used_or_new'].replace(condition_map)

In [22]:
cars['used_or_new'].value_counts()

used    45714
new      3996
demo     3945
Name: used_or_new, dtype: int64

In [23]:
# Rename for convinience 
cars = cars.rename(columns={'used_or_new':'condition'})

#### Status

In [24]:
cars['status'].value_counts()

 Used            16299
 Demo              387
 New In Stock      357
Name: status, dtype: int64

In [25]:
cars = cars.drop(columns=['status']) # Already have condition column so no need

#### Colour

In [26]:
cars['color'].value_counts()

White      6175
Silver     2452
Grey       2440
Black      1969
Blue       1577
           ... 
BRITISH       1
Toy           1
corris        1
Sunset        1
Smoke         1
Name: color, Length: 235, dtype: int64

In [27]:
cars['color'].unique()

array([None, 'Blue', 'White', 'Red', 'Black', 'Silver', 'Grey', 'Mercury',
       'Brown', 'Green', 'blue', 'Orange', 'Whtie', 'TRUE', 'Gold',
       'Liquid', 'Beige', 'Soul', 'Crystal', 'Harmony', 'Yellow',
       'silver', 'Sky', 'white', 'Metallic', 'Bronze', 'WR', 'SILVER',
       'Estoril', 'Slipstream', 'charcoal', 'red', 'black', 'grey',
       'Perl', 'Charcoal', 'Frozen', 'Polar', 'Meteor', 'Mountain',
       'BLACK', 'Baige', 'selenite', 'Diamond', 'Tidal', 'Oolong',
       'Palladium', 'Fusion', 'Gray', 'Graphite', 'Ivory', 'Dark',
       'Jatoba', 'Greg', 'Forged', 'SIlver', 'Pearl', 'Sting', 'Horizon',
       'Magnetic', 'Maroon', 'Bright', 'Burgundy', 'ICE', 'Magic',
       'Misano', 'Flamenco', 'Deep', 'Clear', 'Emerald', 'Fuji',
       'Navarra', 'Autumn', 'light', 'Granite', 'Mojave', 'WHITE', 'Gun',
       'Reflex', 'Purple', 'Cosmos', 'Mica', 'New', 'Gunmetal', 'diamond',
       'Cloth', 'Heron', 'Pacific', 'Petroleum', 'yellow', 'Alpine',
       'Coverline', 'Dayto

In [28]:
# Standardize
cars['color'] = cars['color'].str.strip().str.lower()

# Fix typos
color_fix = {
    'whtie': 'white',
    'whte': 'white',
    'while': 'white',
    'waite': 'white',

    'sliver': 'silver',
    'silber': 'silver',

    'baige': 'beige',

    'charcol': 'charcoal',
    'metalic': 'metallic'
}

cars['color'] = cars['color'].replace(color_fix)

# Detect colours by keywords
def simplify_color(c):
    
    if c is None:
        return None
    
    if 'white' in c or 'snow' in c or 'crystal' in c:
        return 'white'
    
    if 'black' in c or 'ebony' in c:
        return 'black'
    
    if 'silver' in c or 'silber' in c or 'platinum' in c:
        return 'silver'
    
    if 'grey' in c or 'gray' in c or 'charcoal' in c or 'graphite' in c:
        return 'grey'
    
    if 'blue' in c or 'navy' in c or 'sky' in c:
        return 'blue'
    
    if 'red' in c or 'burgundy' in c or 'maroon' in c or 'ruby' in c:
        return 'red'
    
    if 'green' in c or 'emerald' in c:
        return 'green'
    
    if 'yellow' in c:
        return 'yellow'
    
    if 'orange' in c:
        return 'orange'
    
    if 'brown' in c:
        return 'brown'
    
    if 'gold' in c:
        return 'gold'
    
    if 'beige' in c or 'cream' in c or 'ivory' in c:
        return 'beige'
    
    return 'other'

cars['color'] = cars['color'].apply(simplify_color)

In [29]:
cars['color'].value_counts()

white     6272
grey      2495
silver    2489
black     1998
blue      1595
red       1190
other      302
gold       172
orange     155
green      141
brown      138
yellow      72
beige       24
Name: color, dtype: int64

#### Location and States

In [30]:
cars[['location','state']].head()

,location,state
0,Queensland,None
1,Victoria,None
2,Victoria,None
3,Victoria,None
4,New South Wales,None


In [31]:
cars[['location','state']].isna().sum()

location    17043
state       37630
dtype: int64

In [32]:
# Standardize
cars['location'] = cars['location'].str.strip().str.lower()
cars['state'] = cars['state'].str.strip().str.lower()

# Extract state from string
cars['state_from_location'] = cars['location'].str.split(',').str[-1].str.strip()

# Fill missing state using location
cars['state'] = cars['state'].fillna(cars['state_from_location'])

# Remove helper column
cars = cars.drop(columns=['state_from_location'])

In [37]:
cars['state'].value_counts()

nsw                             6285
queensland                      6015
victoria                        4607
western australia               4387
vic                             3808
new south wales                 3396
qld                             2683
wa                              1782
tasmania                         811
sa                               795
south australia                  554
australian capital territory     466
act                              378
n/a                              311
tas                              214
nt                                80
northern territory                40
Name: state, dtype: int64

In [38]:
# Standardize and map
cars['state'] = cars['state'].str.strip().str.lower()

state_map = {
    'new south wales': 'nsw',
    'nsw': 'nsw',

    'victoria': 'vic',
    'vic': 'vic',

    'queensland': 'qld',
    'qld': 'qld',

    'western australia': 'wa',
    'wa': 'wa',

    'south australia': 'sa',
    'sa': 'sa',

    'tasmania': 'tas',
    'tas': 'tas',

    'australian capital territory': 'act',
    'act': 'act',

    'northern territory': 'nt',
    'nt': 'nt',

    'n/a': 'unknown'
}

cars['state'] = cars['state'].replace(state_map)

cars['state'] = cars['state'].fillna('unknown')

In [39]:
cars['state'].value_counts()

unknown    17354
nsw         9681
qld         8698
vic         8415
wa          6169
sa          1349
tas         1025
act          844
nt           120
Name: state, dtype: int64

In [40]:
cars['location'].value_counts()

queensland             6015
victoria               4607
western australia      4387
new south wales        3396
tasmania                811
                       ... 
east toowoomba, qld       1
norwood, sa               1
coolum beach, qld         1
blackburn, vic            1
wongan hills, wa          1
Name: location, Length: 619, dtype: int64

In [41]:
# 
cars = cars.drop(columns=['location'])

#### Low value columns

In [42]:
# Each data row has model + make, so series, title and listing_id don't help
cars = cars.drop(columns=['title', 'series', 'listing_id'])

#### Car make

In [43]:
cars['make'].unique()

array(['alfa', 'aston', 'audi', 'bmw', 'chery', 'chevrolet', 'chrysler',
       'citroen', 'cupra', 'daihatsu', 'dodge', 'fiat', 'ford', 'fpv',
       'genesis', 'great', 'gwm', 'haval', 'holden', 'honda', 'hsv',
       'hyundai', 'isuzu', 'jac', 'jaguar', 'jeep', 'kgm', 'kia', 'land',
       'land rover', 'ldv', 'lexus', 'mahindra', 'mazda', 'mercedes-benz',
       'mg', 'mini', 'mitsubishi', 'nissan', 'opel', 'peugeot', 'porsche',
       'ram', 'renault', 'saab', 'skoda', 'smart', 'ssangyong', 'subaru',
       'suzuki', 'tesla', 'toyota', 'volkswagen', 'volvo', 'great wall',
       'aston martin', 'abarth', 'alfa romeo', 'lamborghini', 'foton',
       'infiniti', 'bentley', 'maserati', 'mclaren', 'byd', 'iveco',
       'rover', 'geely', 'hummer', 'ferrari', 'ineos', 'tata', 'daewoo',
       'hino', 'proton', 'rolls-royce', 'polestar', 'mitsubishi fuso'],
      dtype=object)

In [46]:
cars['make'] = cars['make'].str.strip().str.lower()

make_map = {
    'aston': 'aston martin',
    'great': 'great wall',
    'gwm': 'great wall',
    'land': 'land rover',
    'mercedes-benz': 'mercedes',
    'alfa': 'alfa romeo'
}

cars['make'] = cars['make'].replace(make_map)

In [47]:
sorted(cars['make'].unique())

['abarth',
 'alfa romeo',
 'aston martin',
 'audi',
 'bentley',
 'bmw',
 'byd',
 'chery',
 'chevrolet',
 'chrysler',
 'citroen',
 'cupra',
 'daewoo',
 'daihatsu',
 'dodge',
 'ferrari',
 'fiat',
 'ford',
 'foton',
 'fpv',
 'geely',
 'genesis',
 'great wall',
 'haval',
 'hino',
 'holden',
 'honda',
 'hsv',
 'hummer',
 'hyundai',
 'ineos',
 'infiniti',
 'isuzu',
 'iveco',
 'jac',
 'jaguar',
 'jeep',
 'kgm',
 'kia',
 'lamborghini',
 'land rover',
 'ldv',
 'lexus',
 'mahindra',
 'maserati',
 'mazda',
 'mclaren',
 'mercedes',
 'mg',
 'mini',
 'mitsubishi',
 'mitsubishi fuso',
 'nissan',
 'opel',
 'peugeot',
 'polestar',
 'porsche',
 'proton',
 'ram',
 'renault',
 'rolls-royce',
 'rover',
 'saab',
 'skoda',
 'smart',
 'ssangyong',
 'subaru',
 'suzuki',
 'tata',
 'tesla',
 'toyota',
 'volkswagen',
 'volvo']

#### Car model

### Handling missing values

#### Check for duplicates

In [34]:
cars[['make','model']].duplicated().sum()

52585

In [35]:
cars.duplicated(subset=['make','model','year','kms','price']).sum()

111

In [36]:
cars = cars.drop_duplicates(subset=['make','model','year','kms','price'])

#### Check how many missing

In [37]:
cars.isna().sum().sort_values(ascending=False)

drive_type      37526
seats           36605
listing_id      36605
color           36605
series          36605
variant         36605
cc              22206
title           20587
engine_l         2442
body_type        1182
fuel_type         538
transmission      239
price               0
location            0
state               0
condition           0
kms                 0
year                0
model               0
make                0
source              0
dtype: int64